### DISCUSSION TITLE
物品在玩家/生物间亲手给予（传递）

### DISCUSSION THREAD

**Discussion by BetaFoprhoton** [mouse0w0/MinecraftDeveloperGuide#86] [想法 / Ideas]
## 简要描述：（键盘快捷键、方位待定）
1. 当玩家按住 **左alt键** 时，呼出给予菜单，菜单包含一定距离内玩家名称。
2. 当玩家按住 **左alt键** 后，按 **V键** 在菜单中切换选中将传递给的玩家。（前两步骤参考原版F3 + F4模式切换菜单）
3. 当玩家松开 **左alt键** 时，玩家物品栏选中的物品（主手）格出现传递标记（标记参考create的工具箱物品自动补货）。同时想传递物品玩家屏幕上出现 **等待传递标记** ，包含**被传递玩家头像、名称，选中栏物品栈信息，间隔距离**等信息。
4. 当玩家请求传递时，被请求传递的玩家屏幕上出现包含 **请求传递玩家头像、名称，请求传递的物品栈信息，间隔距离** 等信息。此时被请求传递的玩家可 **轻触左alt** 同意或 **轻触右alt** 拒绝。
5. 当被传递玩家**拒绝**时，传递请求取消，两位玩家屏幕上 **等待传递标记** 显示 **×** 并在几秒后消失 。
6. 当被传递玩家**同意**时，两位玩家屏幕上 **等待传递标记** 添加显示 **√** 并出现另一位玩家相对于这位玩家视线方向指示（<、| 与 >）
7. 当两位玩家之间距离小于2m **&&** 传递玩家手持被传递物品 **&&** 被传递玩家主副手均无物品 时，传递开始，第一、三人称出现传递动画，传递物品玩家 将 被传递物品 **递给** 被传递玩家， **等待传递标记** 消失。
8. 当两位玩家之间距离在2m≤x≤**一定距离**时 **&&** 传递玩家手持被传递物品 **&&** 被传递玩家主副手均无物品 **&&** 传递路线上无遮挡 时，抛物传递开始，第一、三人称出现传递动画，传递物品玩家 将 被传递物品 **抛给** 被传递玩家， **等待传递标记** 消失。
#### 关于传递：
1. 请求传递时，请求传递的玩家不必一直手持该物品。
2. 请求传递时，被选中物品栏的物品数量可减少 / 增加，但是数量为0会取消传递。
3. 当被传递玩家传递过程afk且没有对请求做出拒绝或同意的反应时，传递玩家可走到被传递玩家跟前，将物品 **硬塞** 给被传递玩家。
4. 当玩家呼出传递菜单时，按住 **B** 键，进入要求模式，要求被玩家传递物品给这位玩家。
5. 当请求传递玩家呼出传递菜单时，首次传递按照玩家距离排序，非首次传递当前视线内的玩家排在第一位。如果视线内无玩家，上次传递玩家排第一位。
6. 待完善 **分发（一个玩家分发一个物品以一定数量给多个玩家）、交易（一手交钱，一手拿货）** 等 细节。
7. 上文传递过程拓展到非玩家生物（？

## 技术要点：
1. 键位识别且可配置
2. 菜单gui以及玩家列表实现
3. 等待传递标记实现
4. 传递过程动画实现
5. 抛物实现

## 意义：
1. 原版+ 的轻量化，适用于多种服务器的多种应用场景，有望被收编。
2. 告别玩家间 “掷物于地” 的传递方式，使玩家间传递更安全。
3. 增进玩家游戏体验。

## 实现：
1. 开团队repo，由愿意加入的开发者共同维护。
2. 放到mouse0w0的项目粮仓中，由愿意的开发者pr维护。
3. 个人实现。
（即将开学，没有太多时间完成项目）

**Comment by BetaFoprhoton** [comment]
主要功能非交易，而是强调玩家与玩家间，玩家与生物间**相互递物品**的过程，以及玩家方便快捷的键位，与轻量化的功能

**Comment by Leawind** [comment]
这个想法很酷耶，但我觉得还可以再轻量化一点。

以下是我的想象：

## 玩家A传递物品给玩家B的步骤

1. A 手持要传递的物品
2. A 按下特定按键，A 进入“准备递予物品”状态。用特定姿势来表示 A 的这种状态（比如把物品举过头顶并挥动），这样其他玩家可以看出 A 进入了该状态。
3. 三种传递方式（传递时应播放相应动画）
    1. A 对 B 交互，可以将物品放到B的手上或物品栏里。
    2. B 对 A 交互，即可交换 B 和 A 手中的物品。
    3. 投掷物品。如果 A 与 B 距离过远，无法直接交互，A 可以对着空气按下交互键抛出物品。

## 主要考虑的原则

* 实现功能所需的按键尽可能少，减少按键冲突
* 逻辑尽量符合直觉，降低玩家的学习成本
* 尊重原版特性，玩家可以被动获得物品，但不能在活着时被其他玩家强行取走物品。

**Reply by BetaFoprhoton** [reply]
非常赞同你的想法，确实可以更轻量一点。
上述ui主要参考create的机器设置，但是我想的确实不太简便。
也许可以对远处的B挥动物品，近处朝向B举起物品。如果A正忙，也许可以将待传递的物品换到副手。
还有一些功能需要完善，比如“分发”，怎么让A将物品分发给BCDE等人，我想的是在上述基础上使用鼠标滚轮。也可以学习create的机器设定界面，比如按住传递键呼出菜单，里面可以用鼠标滑动选定传递、分发、收集、请求（物品）等模式，但也许这是后面更多功能的事，现在感觉可以尝试写写看）
